File: frontend/projects/irrtirer/src/app/shared/mosaic-generator/orderer/section-layout/specimen-selection.ts
import { RandomHelper } from '../../../../core/helpers/random-helper';
import { PossibleLayout } from './possible-layout';

export class SpecimenSelection {
    public turnamentSelection(population: PossibleLayout[], k: number): PossibleLayout {
        k = Math.min(k, population.length);

        const randomedIndexes: Set<number> = new Set<number>();

        let randomedIndex: number = RandomHelper.nextInt(0, population.length-1);
        let bestSpecimen: PossibleLayout = population[randomedIndex];
        randomedIndexes.add(randomedIndex);

        while (k > 1) {
            randomedIndex = RandomHelper.nextInt(0, population.length-1);
            if (!randomedIndexes.has(randomedIndex)) {
                --k;
                const selctedSpecimen: PossibleLayout = population[randomedIndex];
                randomedIndexes.add(randomedIndex);

                if (selctedSpecimen.rate > bestSpecimen.rate) {
                    bestSpecimen = selctedSpecimen;
                }
            }
        }

        return bestSpecimen;
    }

    public selectBestInPopulation(population: PossibleLayout[]): PossibleLayout {
        let bestSpecimen: PossibleLayout = population[0];
        for (let i = 1; i < population.length; ++i) {
            if (population[i].rate > bestSpecimen.rate) {
                bestSpecimen = population[i];
            }
        }

        return bestSpecimen;
    }
}

File: backend/Irrtirer/Irrtirer.Generator/Orderer/SectionLayout/SpecimenSelection.cs
using System;
using System.Collections.Generic;

namespace Irrtirer.Generator.Orderer.SectionLayout
{
    internal class SpecimenSelection
    {
        private readonly Random random;

        public SpecimenSelection(Random random)
        {
            this.random = random;
        }

        public PossibleLayout TournamentSelection(PossibleLayout[] population, int k)
        {
            k = Math.Min(k, population.Length);

            HashSet<int> drawnIndexes = new HashSet<int>();

            int drawnIndex = random.Next(0, population.Length);
            PossibleLayout bestSpecimen = population[drawnIndex];
            drawnIndexes.Add(drawnIndex);

            while (k > 1)
            {
                drawnIndex = random.Next(0, population.Length);
                if (!drawnIndexes.Contains(drawnIndex))
                {
                    --k;
                    PossibleLayout selectedSpecimen = population[drawnIndex];
                    drawnIndexes.Add(drawnIndex);

                    if (selectedSpecimen.Rate > bestSpecimen.Rate)
                    {
                        bestSpecimen = selectedSpecimen;
                    }
                }
            }

            return bestSpecimen;
        }

        public PossibleLayout SelectBestInPopulation(PossibleLayout[] population)
        {
            PossibleLayout bestSpecimen = population[0];
            for (int i = 1; i < population.Length; ++i)
            {
                if (population[i].Rate > bestSpecimen.Rate)
                {
                    bestSpecimen = population[i];
                }
            }

            return bestSpecimen;
        }
    }
}
